fvm/transaction.go

```diff
@@ -70,65 +76,113 @@ func (i *TransactionInvocator) Process(
 	proc *TransactionProcedure,
 	st *state.State,
 ) error {
-	env, err := newEnvironment(ctx, vm, st)
-	if err != nil {
-		return err
+
+	var err error
+	var env *hostEnv
+	var blockHeight uint64
+	if ctx.BlockHeader != nil {
+		blockHeight = ctx.BlockHeader.Height
 	}
-	env.setTransaction(proc.Transaction, proc.TxIndex)
 
-	location := common.TransactionLocation(proc.ID[:])
+	numberOfRetries := 0
+	for numberOfRetries = 0; numberOfRetries < int(ctx.MaxNumOfTxRetries); numberOfRetries++ {
+		env, err = newEnvironment(ctx, vm, st)
+		// env construction error is fatal
+		if err != nil {
+			return err
+		}
+		env.setTransaction(proc.Transaction, proc.TxIndex)
+
+		location := common.TransactionLocation(proc.ID[:])
+
+		err = vm.Runtime.ExecuteTransaction(
+			runtime.Script{
+				Source:    proc.Transaction.Script,
+				Arguments: proc.Transaction.Arguments,
+			},
+			runtime.Context{
+				Interface: env,
+				Location:  location,
+			},
+		)
+
+		// break the loop
+		if !i.requiresRetry(err, proc) {
+			break
+		}
+
+		i.logger.Warn().
+			Str("txHash", proc.ID.String()).
+			Uint64("blockHeight", blockHeight).
+			Int("retries_count", numberOfRetries).
+			Uint64("ledger_interaction_used", st.InteractionUsed()).
+			Msg("retrying transaction execution")
+
+		// reset error part of proc
+		// Warning right now the tx requires retry logic doesn't change
+		// anything on state but we might want to revert the state changes (or not commiting)
+		// if we decided to expand it furthur.
+		proc.Err = nil
+		proc.Logs = make([]string, 0)
+		proc.Events = make([]flow.Event, 0)
+		proc.ServiceEvents = make([]flow.Event, 0)
+		proc.Retried++
+	}
 
-	err = vm.Runtime.ExecuteTransaction(
-		runtime.Script{
-			Source:    proc.Transaction.Script,
-			Arguments: proc.Transaction.Arguments,
-		},
-		runtime.Context{
-			Interface: env,
-			Location:  location,
-		},
-	)
+	// (for future) panic if we tried several times and still failing
+	// if numberOfTries == maxNumberOfRetries {
+	// 	panic(err)
+	// }
 
 	if err != nil {
-		i.safetyErrorCheck(err)
+		i.logger.Info().
+			Str("txHash", proc.ID.String()).
+			Uint64("blockHeight", blockHeight).
+			Uint64("ledgerInteractionUsed", st.InteractionUsed()).
+			Msg("transaction executed with error")
 		return err
 	}
 
-	i.logger.Info().
-		Str("txHash", proc.ID.String()).
-		Msgf("(%d) ledger interactions used by transaction", st.InteractionUsed())
-
 	proc.Events = env.getEvents()
 	proc.ServiceEvents = env.getServiceEvents()
 	proc.Logs = env.getLogs()
 
+	i.logger.Info().
+		Str("txHash", proc.ID.String()).
+		Uint64("blockHeight", blockHeight).
+		Uint64("ledgerInteractionUsed", st.InteractionUsed()).
+		Int("retried", proc.Retried).
+		Msg("transaction executed with no error")
```

Comment:
a minor suggestion, to log as `transaction executed successfully`, so that when we search log by `error`, this will not be included.

Reply:
I'll update in the next PR

Reply:
added to [this PR ](https://github.com/onflow/flow-go/pull/471)